# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/vm/NativeObject.h
# Commit: 12574c8b209d
# Full Hash: 12574c8b209d51965201890abdcbd40b553e1190
# Author: Jan de Mooij <jdemooij@mozilla.com>
# Date: 2021-03-23 21:43:59
# Regressor Bug: 1696178
# File Overlap Count: 1
# Description:
#   Bug 1696178 part 4 - Add NativeObject::setLastPropertyForNewDataProperty. r=jonco
#   
#   NativeObject::addEnumerableDataProperty is very hot and this is a measurable speedup
#   because it's not as branchy as setLastProperty.
#   
# ==============================================================================

diff -r aa7ea7bda43e -r 12574c8b209d js/src/vm/NativeObject.h
--- a/js/src/vm/NativeObject.h	Tue Mar 23 12:43:56 2021 +0000
+++ b/js/src/vm/NativeObject.h	Tue Mar 23 12:43:56 2021 +0000
@@ -594,6 +594,11 @@
   // with the object's new slot span.
   MOZ_ALWAYS_INLINE bool setLastProperty(JSContext* cx, Shape* shape);
 
+  // Optimized version of setLastProperty for when |shape| is a new data
+  // property and |shape->previous() == lastProperty()|.
+  MOZ_ALWAYS_INLINE bool setLastPropertyForNewDataProperty(JSContext* cx,
+                                                           Shape* shape);
+
   // Newly-created TypedArrays that map a SharedArrayBuffer are
   // marked as shared by giving them an ObjectElements that has the
   // ObjectElements::SHARED_MEMORY flag set.